### PR TITLE
OEP for frontend dependency upgrade

### PR DESCRIPTION
**Pull request by abdullahwaheed**

### CODE REVIEW

**Comment by feanil** [comment]
This is great documentation, I think 2 things need to change: 

1. Make this an ADR on the Frontend Technology Standard OEP(OEP-11)  See https://github.com/openedx/open-edx-proposals/tree/master/oeps/best-practices/oep-0011/decisions for examples.

2. This should cross link to https://docs.openedx.org/en/latest/developers/how-tos/enable-javascript-upgrade-automation.html  and that how-to should be updated so that the info there matches the content here.  I think the mechanics and sample config can live in the how-to and what we need here is just the decision that we want to automate this and what tool to use.

File: oeps/best-practices/oep-0065-bp-frontend-dependencies.rst [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+     - Best Practice
+   * - Created
+     - 2023-07-17
+   * - Review Period
+     - <start - target end dates for review>
+
+Abstract
+********
+
+Proposes best practices for maintaining dependencies on Javscript repositories of Open edX
```
Comment (arbrandes) [inline oeps/best-practices/oep-0065-bp-frontend-dependencies.rst:42]:
```suggestion
Proposes best practices for maintaining dependencies on Javascript repositories of Open edX
```

File: oeps/best-practices/oep-0065-bp-frontend-dependencies.rst [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+Context
+*******
+
+Openedx mico-frontends(MFEs) use javascript. We maintain these repositoriesusing  To upgrade its dependencies through package.json, we can either use `^` or `~` operators to automatically upgrade major and patch versions respectively.
```
Comment (arbrandes) [inline oeps/best-practices/oep-0065-bp-frontend-dependencies.rst:47]:
```suggestion
Openedx micro-frontends(MFEs) use javascript.  We maintain these repositories using Webpack.  To upgrade its dependencies through package.json, we can either use `^` or `~` operators to automatically upgrade major and patch versions respectively. 
```

File: oeps/best-practices/oep-0065-bp-frontend-dependencies.rst [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+-------------------
+In some cases, you may want to exclude certain dependencies from being updated automatically. Use the ignoreDeps configuration in renovate.json to specify packages that Renovate Bot should not update.
+
+Documentation
+-------------
+Renovate Bot PRs have very good documentation of changes with proper linking.
+
+Best Practices
+**************
+We can configure Renovate Bot in our repositories by introducing `renovate.json` file. Here are some best practices we can follow to
```
Comment (arbrandes) [inline oeps/best-practices/oep-0065-bp-frontend-dependencies.rst:92]:
```suggestion
We can configure Renovate Bot in our repositories by introducing `renovate.json` file. Here are some best practices on how to configure it.
```

File: oeps/best-practices/oep-0065-bp-frontend-dependencies.rst [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        "matchUpdateTypes": [
+          "lockFileMaintenance",
+          "minor",
+          "patch",
+          "pin"
+        ],
+        "automerge": true
+      },
+      {
+        "matchPackagePatterns": ["@edx"],
```
Comment (arbrandes) [inline oeps/best-practices/oep-0065-bp-frontend-dependencies.rst:125]: Soon we'll also have "`@openedx`", so I figure we might as well add it to the list!

File: oeps/best-practices/oep-0065-bp-frontend-dependencies.rst [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        "matchUpdateTypes": ["minor", "patch"],
+        "automerge": true
+      }
+    ],
+    "timezone": "America/New_York"
+  }
+
+We can set schedule of whole Renovate Bot or we can set independent schedules.
+We can also set the automerge rules to be satisfy.
```
Comment (arbrandes) [inline oeps/best-practices/oep-0065-bp-frontend-dependencies.rst:134]:
```suggestion
We can also set the automerge rules to be satisfied. 
```

**Review by arbrandes — CHANGES_REQUESTED** [review]
Thanks for this!  This is exactly the kind of decision we should document.

Aside from nit-picking a couple of typos (sorry, I know it's a draft, but can't help myself), I very much like the idea of documenting this.  Like Feanil, though, I think this should go in as an ADR to OEP-11.  I also support his other suggestions.

File: oeps/best-practices/oep-0066-bp-frontend-dependencies.rst [hunk 6]
```diff
@@ -39,12 +39,12 @@ OEP-0065: Frontend Dependencies Update
...
 ********
 
-Proposes best practices for maintaining dependencies on Javscript repositories of Open edX
+Proposes best practices for maintaining dependencies on Javascript repositories of Open edX
 
 Context
 *******
 
-Openedx mico-frontends(MFEs) use javascript. We maintain these repositoriesusing  To upgrade its dependencies through package.json, we can either use `^` or `~` operators to automatically upgrade major and patch versions respectively. 
+Open edX micro-frontends(MFEs) use javascript. We maintain these repositories using  To upgrade its dependencies through package.json, we can either use `^` or `~` operators to automatically upgrade major and patch versions respectively.
```
Comment (arbrandes) [inline oeps/best-practices/oep-0066-bp-frontend-dependencies.rst:47]: Using...?

File: oeps/best-practices/oep-0011/decisions/0007-renovate.rst [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+Status
+******
+
+Pending
+
+Context
+*******
+
+To keep the micro-frontends(MFEs) of Open edX up-to-date, use some tool to autoamte the process.
```
Comment (arbrandes) [inline oeps/best-practices/oep-0011/decisions/0007-renovate.rst:12]:
```suggestion
To keep the micro-frontends(MFEs) of Open edX up-to-date, use some tool to automate the process.
```

**Review by arbrandes — CHANGES_REQUESTED** [review]
I still think that the proposed text in OEP-0066 should live entirely in the renovate ADR.

(The fact that previous OEP-11 ADRs don't contain a lot of text is only because they were moved out from the OEP and there wasn't a lot of detail to port over.)

File: oeps/best-practices/oep-0011/decisions/0007-renovate.rst (hunk 7 above)
Comment (sarina) [inline oeps/best-practices/oep-0011/decisions/0007-renovate.rst:12]:
```suggestion
To keep the micro-frontends(MFEs) of the Open edX codebase up-to-date, use some tool to automate the process.
```

File: oeps/best-practices/oep-0066-bp-frontend-dependencies.rst [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+     - Best Practice
+   * - Created
+     - 2023-07-17
+   * - Review Period
+     - <start - target end dates for review>
+
+Abstract
+********
+
+Proposes best practices for maintaining dependencies on Javascript repositories of Open edX
```
Comment (sarina) [inline oeps/best-practices/oep-0066-bp-frontend-dependencies.rst:42]:
```suggestion
Proposes best practices for maintaining dependencies on Javascript repositories of the Open edX codebase.
```

File: oeps/best-practices/oep-0066-bp-frontend-dependencies.rst [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+It might lead to some complexities since its hard to debug what upgrade actually caused the issue. 
+To overcome this issue, we can use renovate bot. Renovate Bot is a popular GitHub app that automatically keeps your npm dependencies (including the package-lock.json) up-to-date. Once installed, Renovate Bot scans your repository for outdated packages and creates pull requests with updated versions. You can configure the bot's behavior through a renovate.json file in your repository.
+
+
+Specification
+*************
+
+We can configure renovate bot in our github repositories to update npm packages automatically. 
+
+Here are the advantages of using renovate bot:
```
Comment (sarina) [inline oeps/best-practices/oep-0066-bp-frontend-dependencies.rst:57]: Above you say that this document "Proposes best practices for maintaining dependencies on Javascript repositories", yet this section just explains what options are available. Do we want to actually suggest certain values or value ranges as best practices? For example, should renovate run weekly? Monthly? (I don't know if it's appropriate to suggest values, but wanted to flag it)

**Comment by sarina** [comment]
Hi @abdullahwaheed - is there continued energy to keep movement on this OEP?

**Pull request closed without merging**